dev/check_function_signatures.py

```diff
@@ -0,0 +1,210 @@
+import argparse
+import ast
+import os
+import subprocess
+import sys
+from dataclasses import dataclass
+from pathlib import Path
+from typing import Optional
+
+
+def is_github_actions() -> bool:
+    return os.environ.get("GITHUB_ACTIONS") == "true"
+
+
+@dataclass
+class Signature:
+    lineno: int
+    col_offset: int
+    args: list[str]
+
+
+@dataclass
+class Error:
+    file_path: Path
+    line: int
+    column: int
+    function_name: str
+    message: str
+
+    def format(self, github: bool) -> str:
+        if github:
+            return (
+                f"::warning file={self.file_path},line={self.line},"
+                f"col={self.column}::{self.message}"
+            )
+        else:
+            return f"{self.file_path}:{self.line}:{self.column}: {self.message}"
+
+
+class FunctionSignatureExtractor(ast.NodeVisitor):
+    def __init__(self):
+        self.signatures: dict[str, Signature] = {}
+        self.name_stack: list[str] = []
+
+    def visit_ClassDef(self, node: ast.ClassDef) -> None:
+        self.name_stack.append(node.name)
+        self.generic_visit(node)
+        self.name_stack.pop()
+
+    def visit_FunctionDef(self, node: ast.FunctionDef) -> None:
+        # Is this a private function or a function in a private class?
+        # If so, skip it.
+        if any(n.startswith("_") and not n.startswith("__") for n in [*self.name_stack, node.name]):
+            return
+
+        path = ".".join([*self.name_stack, node.name])
+        self.signatures[path] = Signature(
+            lineno=node.lineno,
+            col_offset=node.col_offset,
+            args=[a.arg for a in node.args.posonlyargs + node.args.args],
```

Comment:
We can ignore keyword only arguments. For example, the following change is not breaking:

```python
def f(*, a=1, b=2):
    pass
```

```python
def f(*, a=1, c=3, b=2):
    pass
```